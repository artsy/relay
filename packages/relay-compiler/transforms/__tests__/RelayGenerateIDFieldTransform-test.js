/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+relay
 */

'use strict';

const {buildSchema, GraphQLNonNull, GraphQLID} = require('graphql');

const GraphQLCompilerContext = require('GraphQLCompilerContext');
const GraphQLIRPrinter = require('GraphQLIRPrinter');
const RelayParser = require('RelayParser');
const RelayTestSchema = require('RelayTestSchema');
const {generateTestsFromFixtures} = require('RelayModernTestUtils');

const {
  transform,
  buildSelectionFromFieldDefinition,
  getIDFieldDefinition,
  getNodeIDFieldDefinition,
} = require('RelayGenerateIDFieldTransform');

describe('RelayGenerateIDFieldTransform', () => {
  generateTestsFromFixtures(
    `${__dirname}/fixtures/generate-id-field-transform`,
    text => {
      const ast = RelayParser.parse(RelayTestSchema, text);
      return new GraphQLCompilerContext(RelayTestSchema)
        .addAll(ast)
        .applyTransforms([transform])
        .documents()
        .map(GraphQLIRPrinter.print)
        .join('\n');
    },
  );

  describe('buildSelectionFromFieldDefinition', () => {
    it('returns an aliased selection', () => {
      expect(
        buildSelectionFromFieldDefinition({
          name: 'id',
          type: new GraphQLNonNull(GraphQLID),
        }),
      ).toEqual({
        kind: 'ScalarField',
        alias: '__id',
        args: [],
        directives: [],
        handles: null,
        metadata: null,
        name: 'id',
        type: new GraphQLNonNull(GraphQLID),
      });
    });

    it('returns an unaliased selection when the name matches the alias', () => {
      expect(
        buildSelectionFromFieldDefinition({
          name: '__id',
          type: new GraphQLNonNull(GraphQLID),
        }),
      ).toEqual({
        kind: 'ScalarField',
        alias: null,
        args: [],
        directives: [],
        handles: null,
        metadata: null,
        name: '__id',
        type: new GraphQLNonNull(GraphQLID),
      });
    });
  });

  describe('getNodeIDFieldDefinition()', () => {
    it('returns nothing in case no Node interface exists', () => {
      const schema = buildSchema(`
        interface NotNode {
          id: ID!
        }
      `);
      expect(getNodeIDFieldDefinition(schema)).toEqual(null);
    });

    it('returns a ID field entry of the Node interface', () => {
      const schema = buildSchema(`
        interface Node {
          customNodeID: ID
        }
      `);
      expect(getNodeIDFieldDefinition(schema).name).toEqual('customNodeID');
    });

    it('returns a ID! field entry of the Node interface', () => {
      const schema = buildSchema(`
        interface Node {
          customNodeID: ID!
        }
      `);
      expect(getNodeIDFieldDefinition(schema).name).toEqual('customNodeID');
    });

    it('asserts that Node has a field of type ID(!)', () => {
      const schema = buildSchema(`
        interface Node {
          id: String
        }
      `);
      expect(() => getNodeIDFieldDefinition(schema)).toThrow();
    });

    it('asserts that Node does not have multiple fields of type ID!', () => {
      const schema = buildSchema(`
        interface Node {
          id: ID!
          customNodeID: ID!
        }
      `);
      expect(() => getNodeIDFieldDefinition(schema)).toThrow();
    });
  });

  describe('getIDFieldDefinition()', () => {
    it('always returns the inflected Node ID! field', () => {
      const schema = buildSchema(`
        interface Node {
          customNodeID: ID!
        }
        type Artwork implements Node {
          customNodeID: ID!
        }
        type Artist {
          customNodeID: ID!
        }
      `);
      expect(
        getIDFieldDefinition(schema, schema.getType('Artwork')).name,
      ).toEqual('customNodeID');
      expect(
        getIDFieldDefinition(schema, schema.getType('Artist')).name,
      ).toEqual('customNodeID');
    });

    it('returns that a type has the fallback `id` field', () => {
      const schema = buildSchema(`
        interface Node {
          customNodeID: ID!
        }
        type Artist {
          id: ID!
        }
      `);
      expect(
        getIDFieldDefinition(schema, schema.getType('Artist')).name,
      ).toEqual('id');
    });
  });
});
