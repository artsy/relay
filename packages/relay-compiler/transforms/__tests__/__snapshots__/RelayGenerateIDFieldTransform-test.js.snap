// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelayGenerateIDFieldTransform matches expected output: abstract-plural.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query CommentQuery {
  node(id: "comment") {
    ... on Comment {
      actors {
        __id: id
      }
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
query CommentQuery {
  node(id: "comment") {
    ... on Comment {
      actors {
        __id: id
      }
    }
    __id: id
  }
}

`;

exports[`RelayGenerateIDFieldTransform matches expected output: node-union.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query UnionQuery {
  maybeNode {
    # ... on Node { __id: id } is autogenerated here because Node is a possible
    # type
    ... on NonNode {
      name
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
query UnionQuery {
  maybeNode {
    ... on NonNode {
      name
    }
    ... on Node {
      __id: id
    }
    ... on FakeNode {
      __id: id
    }
  }
}

`;

exports[`RelayGenerateIDFieldTransform matches expected output: non-node-union.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query UnionQuery {
  neverNode {
    # ... on Node { __id: id } is not autogenerated here since no concrete type
    # implements Node
    ... on NonNode {
      name
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
query UnionQuery {
  neverNode {
    ... on NonNode {
      name
    }
    ... on FakeNode {
      __id: id
    }
  }
}

`;

exports[`RelayGenerateIDFieldTransform matches expected output: query-with-fragment-variables.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query NodeQuery($id: ID!) {
  node(id: $id) {
    ... on FeedUnit {
      actor {
        ...ActorFragment
      }
    }
  }
}

query ViewerQuery {
  viewer {
    actor {
      ...ActorFragment
    }
  }
}

fragment ActorFragment on Actor {
  name
  ... on User {
    birthdate {
      day
      month
    }
  }
  __id: id
}

query AliasedIDQuery {
  viewer {
    actor {
      _id: id
    }
  }
}

query ConditionalIDQuery($cond: Boolean!) {
  viewer {
    actor {
      __id: id @include(if: $cond)
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
query NodeQuery(
  $id: ID!
) {
  node(id: $id) {
    ... on FeedUnit {
      actor {
        ...ActorFragment
        __id: id
      }
    }
    __id: id
  }
}

query ViewerQuery {
  viewer {
    actor {
      ...ActorFragment
      __id: id
    }
  }
}

fragment ActorFragment on Actor {
  name
  ... on User {
    birthdate {
      day
      month
    }
  }
  __id: id
}

query AliasedIDQuery {
  viewer {
    actor {
      _id: id
      __id: id
    }
  }
}

query ConditionalIDQuery(
  $cond: Boolean!
) {
  viewer {
    actor {
      __id: id @include(if: $cond)
      __id: id
    }
  }
}

`;

exports[`RelayGenerateIDFieldTransform matches expected output: reuses-existing-selection.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query StoryQuery {
  story {
    lastName
    __id: id
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
query StoryQuery {
  story {
    lastName
    __id: id
  }
}

`;
